src/MultiplayerMod/Multiplayer/Components/CursorComponent.cs

```diff
@@ -82,19 +84,45 @@ public class CursorComponent : MonoBehaviour {
         if (!initialized)
             return;
 
-        var currentScreen = KScreenManager.Instance.screenStack.FirstOrDefault(screen => screen.mouseOver);
         var otherClientScreen =
             KScreenManager.Instance.screenStack.FirstOrDefault(screen => screen.GetType() == ScreenType);
-        var isOnTheSameScreen = currentScreen?.GetType() == ScreenType;
 
         // If we see a screen where other player is - show cursor within that screen.
         var showScreenOrWorldCursor =
             CursorWithinScreen.CurrentPosition != null && (otherClientScreen?.isActive ?? false);
         gameObject.transform.position = showScreenOrWorldCursor
             ? ScreenToWorld(otherClientScreen!, (Vector3) CursorWithinScreen.CurrentPosition!)
             : camera.WorldToScreenPoint((Vector3) CursorWithinWorld.CurrentPosition!);
-        textComponent.text = PlayerName + (isOnTheSameScreen ? "" : $" ({ScreenName ?? "World"})");
-        imageComponent.color = textComponent.color = isOnTheSameScreen ? Color.white : new Color(1, 1, 1, 0.5f);
+        var screenUnderCursor = FindScreenUnderCursor(gameObject.transform.position);
+        var showScreeName = screenUnderCursor?.GetType() != ScreenType;
+        textComponent.text = PlayerName + (showScreeName ? $" ({ScreenName ?? "World"})" : "");
+        imageComponent.color = textComponent.color = showScreeName ? new Color(1, 1, 1, 0.5f) : Color.white;
+    }
+
+    private KScreen? FindScreenUnderCursor(Vector2 cursor) {
+        var eventData = new PointerEventData(UnityEngine.EventSystems.EventSystem.current) {
+            position = cursor
+        };
+
+        var results = new List<RaycastResult>();
+        UnityEngine.EventSystems.EventSystem.current.RaycastAll(eventData, results);
```

Comment:
Not sure about performance, sounds like it's something heavy.

Reply:
Just tested locally:
- AVG exec time of method is 0.3 ms (entire update is 0.31ms)
- RaycastAll uses only GraphicRaycaster's (no physic raycasting is happining)
- There are ~50 GraphicRaycaster in total in the game (various screen, canvases, buttons)
- Looking into KLEI code itslef they also seems to use this from time to time in various components.

Being said I would suggest to leave it as is (for now).